### PR TITLE
Vgg features

### PR DESCRIPTION
**Pull request by zaccharieramzi**
For now it's just the requirements and the image pre-processing (even though it's very very slow : ~ 13 min...).

### CODE REVIEW

File: image_processing.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+            - ndarray (n, 244, 244, 3): the input data (x_train), images
+            - list (n): the labels of the images (y_train)
+    '''
+    x_train = list()
+    y_train = list()
+    # initialize COCO api for instance annotations
+    coco = COCO(ann_file)
+    # load all categories
+    cats = coco.loadCats(coco.getCatIds())
+    nms = [cat['name'] for cat in cats]
```
Comment (louismartin) [inline image_processing.py]: Why not names instead of nms ?

File: image_processing.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    for cat_id in cat_ids:
+        # Get all the image ids related to that category
+        img_ids = set(coco.catToImgs[cat_id])
+        for img_id in img_ids:
+
+            if not(img_id in registered_img_ids):
+                # for each image we haven't processed yet, we process it
+                img = coco.loadImgs(img_id)[0]
+                img_mat = misc.imread('/'.join(
+                    [data_dir, data_type, img['file_name']]))
```
Comment (louismartin) [inline image_processing.py]: For joining pathes, a robust approach is to use:
import os.path as op
path = op.join('foo', 'bar', 'file.txt')

File: image_processing.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+                # We need to crop image on both sides in order to have
+                # standardized images in terms of shape
+                dimensions = img_mat.shape
+                if len(dimensions) == 3:
+                    tic = time.time()
+                    bigger_dim = np.argmax(dimensions[:2])
```
Comment (louismartin) [inline image_processing.py]: I would put a small comment to say that this part is for cropping

File: image_processing.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+                        right_crop = math.ceil(difference / 2)
+                        if bigger_dim == 0:
+                            x = img_mat[left_crop:-right_crop, :, :]
+                        else:
+                            x = img_mat[:, left_crop:-right_crop, :]
+                    else:
+                        x = img_mat
+                    x = misc.imresize(x, output_shape)
+                    x_train = x_train, x
+                    y_train = y_train, coco.loadCats(cat_id)[0]['name']
```
Comment (louismartin) [inline image_processing.py]: We won't actually need the labels at first (only needed for training the model, not for CCA)

File: image_processing.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+output_shape = (244, 244, 3)
+
+
+def load_images():
+    '''Preprocessing of the MS Coco dataset
+        Output:
+            - ndarray (n, 244, 244, 3): the input data (x_train), images
+            - list (n): the labels of the images (y_train)
+    '''
+    x_train = list()
```
Comment (louismartin) [inline image_processing.py]: Maybe it's faster to instantiate x_train with np.zeros(...) as we know the final size already. This way we allocate all memory at once.

I would have written X_train (capital X) because it is a matrix.

File: vgg.py [hunk 6]
```diff
@@ -0,0 +1,4 @@
+import numpy as np
+from keras.applications import vgg19
+
+net = vgg19.VGG19()
```
Comment (louismartin) [inline vgg.py:7]: Easy $$$

**Review by louismartin — COMMENTED** [review]
Looks good to me overall !

File: image_processing.py (hunk 1 above)
Comment (zaccharieramzi) [reply image_processing.py]: Ils le faisaient comme ça dans la démo de l'API mais je vais changer ce sera plus clair. Même en `cat_names`.

File: image_processing.py (hunk 5 above)
Comment (zaccharieramzi) [reply image_processing.py]: Pour l'instant je ne connais tjs pas le nombre d'images utilisables. Effectivement à la fin ce sera sûrement plus avantageux de faire comme ça.

J'aime pas trop utiliser des majuscules mais je pense que tu as raison : https://www.python.org/dev/peps/pep-0008/#id34

File: image_processing.py (hunk 2 above)
Comment (zaccharieramzi) [reply image_processing.py]: Agreed.

File: image_processing.py (hunk 3 above)
Comment (zaccharieramzi) [reply image_processing.py]: +1

File: image_processing.py (hunk 4 above)
Comment (zaccharieramzi) [reply image_processing.py]: À la lumière de ce que tu as dit, je suis d'accord.
Cela dit, ça me fait penser qu'il faudra se mettre d'accord sur un ordre pour les données, du coup vois si ça te va l'ordre que j'ai choisi, pour que l'intégration soit pas trop galère.

Comment (louismartin) [reply image_processing.py]: Ouai et puis d'un côté ça pourra nous servir de ground truth pour la visualisation ou pour la third view.

File: image_processing.py (hunk 3 above)
Comment (zaccharieramzi) [reply image_processing.py]: En fait je l'avais déjà mis un peu au dessus non?

File: image_processing.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+import numpy as np
+from pycocotools.coco import COCO
+from scipy import misc
+from tqdm import tqdm_notebook
+
+data_dir = "dataset"
+data_type = "train2014"
+ann_file = '{0}/annotations/instances_{1}.json'.format(data_dir, data_type)
+ann_file = op.join(
+    data_dir, "annotations", "instances_{0}.json".format(data_type))
```
Comment (louismartin) [inline image_processing.py:13]: Supprimer la ligne 12 du coup ?

File: image_processing.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+
+def process_image(img):
+    '''From an image given by the MS Coco API, returns the ndarray standardized
+    if the image is tractable, empty list o.w.
+        Args:
+            - img (dict): the image dictionary returned by the MS Coco API
+        Output:
+            - ndarray: the matrix associated with the cropped and resized image
+    '''
```
Comment (louismartin) [inline image_processing.py:25]: En général on passe directement la ndarray de l'image a une fonction de processing, elle ne doit être dépendante de la façon dont on doit lire l'image (genre le jour ou tu changes la clé du dict, faudra changer la fonction du coup).

File: image_processing.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+
+def load_images(categories=None):
+    '''Preprocessing of the MS Coco dataset
+        Output:
+            - ndarray (n, 244, 244, 3): the input data (X_train), images
+            - list (n): the labels of the images (Y_train)
+    '''
+    X_train = list()
+    Y_train = list()
```
Comment (louismartin) [inline image_processing.py:60]: Marrant, moi j'instancie tjs les listes par [], ça a un interet de faire list() ?

File: image_processing.py [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    # initialize COCO api for instance annotations
+    coco = COCO(ann_file)
+    if categories:
+        names = categories
```
Comment (louismartin) [inline image_processing.py:66]: Peut-être mettre un petit comment pour que le lecteur se rappelle que la on prend que un subset des catégories

File: image_processing.py [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+    else:
+        return []
+
+
+def load_images(categories=None):
+    '''Preprocessing of the MS Coco dataset
+        Output:
+            - ndarray (n, 244, 244, 3): the input data (X_train), images
+            - list (n): the labels of the images (Y_train)
+    '''
```
Comment (louismartin) [inline image_processing.py:58]: Il manque l'input

File: image_processing.py [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+            if not(img_id in registered_img_ids):
+                # for each image we haven't processed yet, we process it
+                img = coco.loadImgs(img_id)[0]
+                X = process_image(img)
+                if len(X) > 0:
+                    X_train = X_train + [X]
+                    Y_train = Y_train + [cat_id]
+        registered_img_ids = img_ids.union(img_ids)
+    X_train = np.array(X_train)
+    return X_train, Y_train
```
Comment (louismartin) [inline image_processing.py]: Est-ce qu'on pourrait retourner une liste contenant les ids des images utilisées ?
Comme ça je pourrais réutiliser les même pour les textual embeddings.

File: vgg.py [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+from keras.optimizers import SGD
+import numpy as np
+
+
+net = vgg19.VGG19()
+sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
+
+
+def compile_net():
+    net.compile(optimizer=sgd, loss='categorical_crossentropy')
```
Comment (louismartin) [inline vgg.py:14]: Je suis pas fan des fonctions qui accèdent à des variables globales comme ça. Est-ce que ça a vraiment une utilité d'écrire cette fonction pour une seule ligne ? (ça peut rendre le code opaque en plus, parce que l'utilisateur ne voit pas direct la loss et l'optimizer utilisés)

File: vgg.py [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+
+
+def compute_vgg_features(X, layer=2):
+    f = K.function([net.layers[0].input, K.learning_phase()],
+                   [net.layers[-layer].output])
+    return f([im, 0])[0]
```
Comment (louismartin) [inline vgg.py]: Tu utilises im en sortie alors qu'on a passé X en entrée, erreur? La variable a changé de nom entre temps ?

File: visual_features.ipynb [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "Using TensorFlow backend.\n"
+     ]
```
Comment (louismartin) [inline visual_features.ipynb]: Oublie pas de faire "Cell -> All outputs -> Clear all" avant de commit un notebook pour faciliter le versioning et la lecture.

**Review by louismartin — COMMENTED** [review]
Ok pour moi.

File: visual_features.ipynb (hunk 15 above)
Comment (zaccharieramzi) [reply visual_features.ipynb]: Tain j'étais sûr de l'avoir fait...

File: vgg.py (hunk 14 above)
Comment (zaccharieramzi) [reply vgg.py]: yes erreur

File: vgg.py (hunk 13 above)
Comment (zaccharieramzi) [reply vgg.py:14]: bah l'idée c'était de ne pas avoir à compiler le réseau de neurones quand on importe le fichier.
Si tu veux je peux ne pas mettre de fonction.

File: image_processing.py (hunk 12 above)
Comment (zaccharieramzi) [reply image_processing.py]: yes agreed

File: image_processing.py (hunk 9 above)
Comment (zaccharieramzi) [reply image_processing.py:60]: je sais pas trop. Là je voulais que ce soit clair (genre extra-clair) que tous ces éléments sont des listes.

Comment (zaccharieramzi) [reply image_processing.py:60]: http://stackoverflow.com/questions/2972212/creating-an-empty-list-in-python Je trouve que c'est une très bonne réponse. Donc ici je vais garder `list()`.

File: image_processing.py (hunk 7 above)
Comment (zaccharieramzi) [reply image_processing.py:13]: oui évidemment

File: image_processing.py (hunk 8 above)
Comment (zaccharieramzi) [reply image_processing.py:25]: hmm ouais mais ici ce sera soit dans cette fonction soit dans l'autre. Là je préférais me dire que ça faisait partie du processing de savoir lire l'adresse de l'image.

File: image_processing.py (hunk 11 above)
Comment (zaccharieramzi) [reply image_processing.py:58]: yep

File: image_processing.py (hunk 10 above)
Comment (zaccharieramzi) [reply image_processing.py:66]: d'ac
